src/plugins/fs/udf.c

```diff
@@ -0,0 +1,503 @@
+            return FALSE;
+
+        args[2] = g_strdup_printf ("--blocksize=%d", detected_bs);
+    }
+
+    if (media_type)
+        args[3] = g_strdup_printf ("--media-type=%s", media_type);
+    else
+        args[3] = g_strdup ("--media-type=hd");
+
+    if (revision)
+        args[4] = g_strdup_printf ("--udfrev=%s", revision);
+    else
+        args[4] = g_strdup ("--udfrev=0x201");
+
+    ret = bd_utils_exec_and_report_error (args, extra, error);
+
+    g_free ((gchar *) args[2]);
+    g_free ((gchar *) args[3]);
+    g_free ((gchar *) args[4]);
+
+    return ret;
+}
+
+/**
+ * bd_fs_udf_wipe:
+ * @device: the device to wipe a UDF signature from
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether the UDF signature was successfully wiped from the @device or
+ *          not
+ *
+ * Tech category: %BD_FS_TECH_UDF-%BD_FS_TECH_MODE_WIPE
+ */
+gboolean bd_fs_udf_wipe (const gchar *device, GError **error) {
+    return wipe_fs (device, "udf", FALSE, error);
+}
+
+/**
+ * bd_fs_udf_set_label:
+ * @device: the device containing the file system to set label for
+ * @label: label to set
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether the label of UDF file system on the @device was
+ *          successfully set or not
+ *
+ * Tech category: %BD_FS_TECH_UDF-%BD_FS_TECH_MODE_SET_LABEL
+ */
+gboolean bd_fs_udf_set_label (const gchar *device, const gchar *label, GError **error) {
+    const gchar *args[6] = {"udflabel", "--utf8", NULL, NULL, device, NULL};
+    g_autofree gchar *vid = NULL;
+    gboolean ret = FALSE;
+
+    if (!check_deps (&avail_deps, DEPS_UDFLABEL_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    /* there are different limits for lvid and vid and we need to set both so lets just
+       truncate vid (GParted does the same thing) */
+    if (g_str_is_ascii (label) && strlen (label) > 30)
+        vid = g_strndup (label, 30);
+    else if (g_utf8_validate (label, -1, NULL) && g_utf8_strlen (label, -1) > 15)
+        vid = g_utf8_substring (label, 0, 15);
+    else
+        vid = g_strdup (label);
+
+    args[2] = g_strdup_printf ("--lvid=%s", label);
+    args[3] = g_strdup_printf ("--vid=%s", vid);
+
+    ret = bd_utils_exec_and_report_error (args, NULL, error);
+
+    g_free ((gchar *) args[2]);
+    g_free ((gchar *) args[3]);
+
+    return ret;
+}
+
+/**
+ * bd_fs_udf_check_label:
+ * @label: label to check
+ * @error: (out) (allow-none): place to store error
+ *
+ * Returns: whether @label is a valid label for the UDF file system or not
+ *          (reason is provided in @error)
+ *
+ * Tech category: always available
+ */
+gboolean bd_fs_udf_check_label (const gchar *label, GError **error) {
+    if (g_str_is_ascii (label)) {
+        if (strlen (label) > 126) {
+            g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_LABEL_INVALID,
+                        "Label for UDF filesystem must be at most 126 ASCII characters long.");
```

Comment:
Why _at most 126 ASCII characters_? In UDF label can contain at most 126 UNICODE code points if all code points are up to the U+FF. Or 63 UNICODE code points if there is at least one code point above U+FF.

Reply:
Oh, I can't read :facepalm: I read it as 126 ascii and 63 unicode. Ok, back to the drawing board.